### PR TITLE
refine scheduler service component

### PR DESCRIPTION
**Pull request by hongyh13**
1. Refine the scheduler service component.
2. Add more unit tests for scheduler service.
3. Refine metadata store.

### CODE REVIEW

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 1]
```diff
@@ -64,188 +73,260 @@ public class SchedulerService : ISchedulerService
...
             {
-                var delayTask = Task.Delay(TimeSpan.FromSeconds(SchedulerServicePullingIntervalInSeconds), CancellationToken.None);
+                // delay task should stop without throwing exception when is cancellation requested.
+                Task delayTask = Task.Delay(TimeSpan.FromSeconds(SchedulerServicePullingIntervalInSeconds), cancellationToken).ContinueWith(_ => { });
 
                 try
                 {
-                    // try next time if the queue client isn't initialized.
+                    // try next time if the storage aren't initialized.
                     if (!_queueClient.IsInitialized() || !_metadataStore.IsInitialized())
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:92]: Does those two initializations also need lock to make them thread-safety?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 2]
```diff
@@ -64,188 +73,260 @@ public class SchedulerService : ISchedulerService
...
-                // this case should not happen.
-                if (currentTriggerEntity != null)
+                try
                 {
-                    switch (currentTriggerEntity.TriggerStatus)
+                    scheduledToEnd = await CheckAndUpdateTriggerEntityAsync(cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    // if cancelled, the exception will be caught here and exit while loop as cancellationToken.IsCancellationRequested is true
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:255]: Seems not exit?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 3]
```diff
@@ -64,188 +73,260 @@ public class SchedulerService : ISchedulerService
...
-                }
-                catch (RequestFailedException ex) when (ex.ErrorCode == AzureStorageErrorCode.GetEntityNotFoundErrorCode)
-                {
-                    await CreateInitialTriggerLeaseEntityAsync(cancellationToken);
-
-                    // try to get trigger lease entity after insert
-                    triggerLeaseEntity = await _metadataStore.GetTriggerLeaseEntityAsync(_queueType, cancellationToken);
-                    needRenewLease = false;
-                }
+            _logger.LogInformation($"Scheduler instance {_instanceGuid} starts to acquires lease.");
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:141]: Fixed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 4]
```diff
@@ -64,188 +73,260 @@ public class SchedulerService : ISchedulerService
...
+        /// <returns>return true if The job has been scheduled to end, otherwise return false</returns>
+        private async Task<bool> ProcessInternalAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Scheduler instance {_instanceGuid} starts internal processing.");
+
+            using CancellationTokenSource runningCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            CancellationToken runningCancellationToken = runningCancellationTokenSource.Token;
+
+            // renew lease task will complete if cancelled or any exception
+            Task renewLeaseTask = LoopToRenewLeaseAsync(runningCancellationToken);
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:194]: Sure. renamed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 5]
```diff
@@ -334,91 +424,142 @@ private async Task CreateNextTriggerAsync(CurrentTriggerEntity currentTriggerEnt
...
+                DateTimeOffset.UtcNow.AddMinutes(-1 * JobConfigurationConstants.JobQueryLatencyInMinutes);
 
-            if (_jobConfiguration?.EndTime != null && endTime > _jobConfiguration.EndTime)
+            if (lastTriggerEndTime != null)
             {
-                endTime = (DateTimeOffset)_jobConfiguration.EndTime;
+                DateTime nextOccurrenceTime =
+                    _crontabSchedule.GetNextOccurrence(((DateTimeOffset)lastTriggerEndTime).DateTime);
+                nextOccurrenceTime = DateTime.SpecifyKind(nextOccurrenceTime, DateTimeKind.Utc);
+                DateTimeOffset nextOccurrenceOffsetTime = nextOccurrenceTime;
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:476]: Fixed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/AzureTableMetadataStore.cs [hunk 6]
```diff
@@ -67,50 +78,55 @@ public bool IsInitialized()
...
+            {
+                await _metadataTableClient.UpdateEntityAsync(
+                    tableEntity,
+                    tableEntity.ETag,
+                    cancellationToken: cancellationToken);
+                return true;
+            }
+            catch (RequestFailedException ex) when (ex.ErrorCode == AzureStorageErrorCode.UpdateEntityPreconditionFailedErrorCode)
+            {
+                _logger.LogInformation("Failed to update entity, the etag is not satisfied.");
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/AzureTableMetadataStore.cs:107]: Fixed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs (hunk 2 above)
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:255]: Yes, if it isn't a cancellation exception, will not exit and retry later.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 7]
```diff
@@ -64,188 +73,260 @@ public class SchedulerService : ISchedulerService
...
+            _diagnosticLogger.LogInformation($"Scheduler instance {_instanceGuid} starts running.");
+            _logger.LogInformation($"Scheduler instance {_instanceGuid} starts running.");
 
-            while (!stopRunning && !serviceCancellationToken.IsCancellationRequested)
+            bool scheduledToEnd = false;
+            while (!scheduledToEnd && !cancellationToken.IsCancellationRequested)
             {
-                var delayTask = Task.Delay(TimeSpan.FromSeconds(SchedulerServicePullingIntervalInSeconds), CancellationToken.None);
+                // delay task should stop without throwing exception when is cancellation requested.
+                Task delayTask = Task.Delay(TimeSpan.FromSeconds(SchedulerServicePullingIntervalInSeconds), cancellationToken).ContinueWith(_ => { });
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:87]: Sure. addressed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 8]
```diff
@@ -334,91 +424,142 @@ private async Task CreateNextTriggerAsync(CurrentTriggerEntity currentTriggerEnt
...
+                        await intervalDelayTask;
                     }
                     else
                     {
-                        triggerLeaseEntity.HeartbeatDateTime = DateTimeOffset.UtcNow;
-
-                        await _metadataStore.UpdateEntityAsync(
-                            triggerLeaseEntity,
-                            cancellationToken: cancellationToken);
+                        stopRenew = true;
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:519]: Addressed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 9]
```diff
@@ -64,188 +73,260 @@ public class SchedulerService : ISchedulerService
...
+            CancellationToken runningCancellationToken = runningCancellationTokenSource.Token;
+
+            // renew lease task will complete if cancelled or any exception
+            Task renewLeaseTask = LoopToRenewLeaseAsync(runningCancellationToken);
+
+            // process trigger task will complete if cancelled or the job has been scheduled to end.
+            Task<bool> processTriggerTask = LoopToCheckAndUpdateTriggerEntityAsync(runningCancellationToken);
+
+            // if one of these two tasks returns, should cancel the other one.
+            await Task.WhenAny(renewLeaseTask, processTriggerTask);
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:200]: Add a log for it.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 10]
```diff
@@ -44,15 +45,23 @@ public class SchedulerService : ISchedulerService
...
+
+            _startTime = jobConfiguration.Value.StartTime;
+            _endTime = jobConfiguration.Value.EndTime;
+
+            if (_startTime != null && _endTime != null && _startTime >= _endTime)
+            {
+                throw new ConfigurationErrorException("The start time should be less than the end time.");
+            }
+
             _crontabSchedule = CrontabSchedule.Parse(jobConfiguration.Value.SchedulerCronExpression, new CrontabSchedule.ParseOptions { IncludingSeconds = true });
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:61]: use lazy to parse Scheduler Cron Expression

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs (hunk 1 above)
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:96]: Per our offline discussion, the design is ok.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs]: Per our offline discussion, add validation on `ConfiguraionValidatorV1` to make sure the `SchedulerCronExpression` is valid.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 11]
```diff
@@ -334,91 +428,142 @@ private async Task CreateNextTriggerAsync(CurrentTriggerEntity currentTriggerEnt
...
+
+            if (triggerLeaseEntity.WorkingInstanceGuid != _instanceGuid)
+            {
+                _logger.LogInformation($"Scheduler instance {_instanceGuid} Fails to renew lease, the retrieved lease trigger entity's working instance {triggerLeaseEntity.WorkingInstanceGuid} doesn't match.");
+                return false;
+            }
+
+            triggerLeaseEntity.HeartbeatDateTime = DateTimeOffset.UtcNow;
+
+            await _metadataStore.TryUpdateEntityAsync(
```
Comment (sowu880) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:561]: Is it possible return false here?

**Review by moria97 — APPROVED** [review]

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs (hunk 11 above)
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs:561]: Yes, it is possible return false, fixed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs [hunk 12]
```diff
@@ -334,91 +428,142 @@ private async Task CreateNextTriggerAsync(CurrentTriggerEntity currentTriggerEnt
...
                     }
-                    else
-                    {
-                        triggerLeaseEntity.HeartbeatDateTime = DateTimeOffset.UtcNow;
 
-                        await _metadataStore.UpdateEntityAsync(
-                            triggerLeaseEntity,
-                            cancellationToken: cancellationToken);
-                    }
+                    await intervalDelayTask;
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs]: Sure, remove break and await delay task at finally.

**Pull request merged**